File: 2018/03_NoMatterHowYouSliceIt/claim.py
# ======================================================================
# No Matter How You Slice It
#   Advent of Code 2018 Day 03 -- Eric Wastl -- https://adventofcode.com
#
# Computer simulation by Dr. Dean Earl Wright III
# ======================================================================

# ======================================================================
#                             c l a i m . p y
# ======================================================================
"Claim for No Matter How You Slice It problem of Advent of Code 2018-03"

# ----------------------------------------------------------------------
#                                                                 import
# ----------------------------------------------------------------------
import re

# ----------------------------------------------------------------------
#                                                              constants
# ----------------------------------------------------------------------
PARSE = re.compile(r'^#([0-9]+) @ ([0-9]+),([0-9]+): ([0-9]+)x([0-9]+)$')

# ======================================================================
#                                                                  Claim
# ======================================================================


class Claim():
    """Object representing a single fabric claim"""

    def __init__(self,                          # pylint: disable=R0913
                 number=0,
                 left=0,
                 top=0,
                 width=1,
                 height=1,
                 text=None):

        # 1. Set the values
        self.number = number
        self.left = left
        self.top = top
        self.width = width
        self.height = height

        # 2. If there is text, decode and save those values
        if text:
            parse = PARSE.match(text)
            self.number = int(parse.group(1))
            self.left = int(parse.group(2))
            self.top = int(parse.group(3))
            self.width = int(parse.group(4))
            self.height = int(parse.group(5))

        # 3. Compute bottom and right
        self.right = self.left + self.width
        if self.width:
            self.right -= 1
        self.bottom = self.top + self.height
        if self.height:
            self.bottom -= 1

    def does_overlap(self, other):
        "Does this claim overlap with the other claim?"

        # 0. Preconditions
        assert other is not None
        assert isinstance(other, Claim)

        # 1. Check column displacements
        if self.right < other.left or self.left > other.right:
            # print("not column %s %d | %s %d" % (str(self), self.right, str(other), other.right))
            return False

        # 2. Check row displacement
        if self.bottom < other.top or self.top > other.bottom:
            # print("not row %s %d | %s %d" % (str(self), self.bottom, str(other), other.bottom))
            return False

        # 3. We do indeed overlap
        return True

    def __str__(self):
        return "#%d @ %d,%d: %dx%d" % (self.number,
                                       self.left,
                                       self.top,
                                       self.width,
                                       self.height)


# ----------------------------------------------------------------------
#                                                  module initialization
# ----------------------------------------------------------------------
if __name__ == '__main__':
    pass

# ======================================================================
# end                        c l a i m . p y                         end
# ======================================================================

File: 2018/03_NoMatterHowYouSliceIt/claim.go
// ======================================================================
// No Matter How You Slice It
//   Advent of Code 2018 Day 03 -- Eric Wastl -- https://adventofcode.com
//
// lua implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                         c l a i m . g o
// ======================================================================
// Claim for the Advent of Code 2018 Day 03 puzzle

package main

// ----------------------------------------------------------------------
//                                                                imports
// ----------------------------------------------------------------------
import (
	"fmt"
	"regexp"
	"strconv"
)

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const ()

var (
	claimExp = regexp.MustCompile("\\d+")
)

// ----------------------------------------------------------------------
//                                                                Min/Max
// ----------------------------------------------------------------------

func Min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func Max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

// ======================================================================
//                                                                  Claim
// ======================================================================

type Claim struct {
	Part2    bool
	Number   int
	Text     string
	Left     int
	Top      int
	Width    int
	Height   int
	Right    int
	Bottom   int
	Conflict bool
}

func NewClaim(part2 bool, text string) *Claim {
	// Claim Object for No Matter How You Slice It

	// 1. Set the initial values
	p := new(Claim)
	p.Part2 = part2
	p.Text = text
	p.Conflict = false

	// 2. Process the text
	parts := claimExp.FindAllString(text, -1)
	if len(parts) != 5 {
		fmt.Printf("regexp failed for line '%s'\n", text)
		return nil
	}

	// 3. Save the parts
	p.Number, _ = strconv.Atoi(parts[0])
	p.Left, _ = strconv.Atoi(parts[1])
	p.Top, _ = strconv.Atoi(parts[2])
	p.Width, _ = strconv.Atoi(parts[3])
	p.Height, _ = strconv.Atoi(parts[4])

	// 4. Compute right and bottom
	p.Right = p.Left + p.Width - 1
	p.Bottom = p.Top + p.Height - 1

	// 5. Return the new clain
	return p
}

func (p *Claim) Overlap(other *Claim) []string {

	// 1. initially no overlap
	var overlaps []string

	// 2. Elimate nonconflicting claims
	//fmt.Printf("Overlap step 2 %d %d: p.Right=%d o.Left=%d p.Bottom=%d o.Top=%d\n",
	//	p.Number, other.Number, p.Right, other.Left, p.Bottom, other.Top)
	if p.Right < other.Left || p.Left > other.Right {
		return overlaps
	}
	if p.Bottom < other.Top || p.Top > other.Bottom {
		return overlaps
	}

	// 3. Determine the range of overlap
	left := Max(p.Left, other.Left)
	right := Min(p.Right, other.Right)
	top := Max(p.Top, other.Top)
	bottom := Min(p.Bottom, other.Bottom)
	//fmt.Printf("Overlap step 3 %d %d: left=%d right=%d top=%d bottom=%d\n",
	//	p.Number, other.Number, left, right, top, bottom)

	// 4. Loop for the overlap seats
	for col := left; col < right+1; col++ {
		for row := top; row < bottom+1; row++ {
			overlaps = append(overlaps, fmt.Sprintf("%dx%d", col, row))
		}
	}

	// 5. Mark the claims in conflict
	p.Conflict = true
	other.Conflict = true

	// 6. Return the overlaps
	//fmt.Printf("First=%s, Last=%s\n",
	//	overlaps[0], overlaps[len(overlaps)-1])
	return overlaps
}

// ======================================================================
// end                      c l a i m . g o                     end
// ======================================================================
